src/main/java/org/opensearch/security/privileges/ResourceAccessEvaluator.java

```diff
@@ -0,0 +1,212 @@
+        this.resourceSharingIndexHandler.fetchSharingInfo(req.index(), req.id(), ActionListener.wrap(document -> {
+            if (document == null) {
+                // TODO check whether we should mark response as not allowed. At present, it just returns incomplete response and hence is
+                // delegated to next evaluator
+                log.debug("No resource sharing record found for resource {} and index {}, skipping evaluation.", req.id(), req.index());
+                latch.countDown();
+                return;
+            }
+
+            // If the document is public, action is allowed
+            // If user is the owner, action is allowed
+            if (document.isSharedWithEveryone() || document.isCreatedBy(user.getName())) {
+                presponse.allowed = true;
+                shouldMarkAsComplete.set(true);
+                String message = document.isSharedWithEveryone()
+                    ? "Publicly shared resource"
+                    : "User " + user.getName() + " is the owner of the resource";
+                log.debug("{} {}, granting access.", message, req.id());
+                latch.countDown();
+                return;
+            }
+
+            // Check whether user or their roles match any access-levels this resource is shared at
+            Set<String> accessLevels = new HashSet<>();
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.USERS, Set.of(user.getName())));
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.ROLES, userRoles));
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.BACKEND_ROLES, userBackendRoles));
+
+            // if no access-levels match, then action is not allowed
+            if (accessLevels.isEmpty()) {
+                presponse.allowed = false;
+                log.debug("Resource {} is not shared with user {}", req.id(), user.getName());
+                shouldMarkAsComplete.set(true);
+                latch.countDown();
+                return;
+            }
+
+            // Expand access-levels and check if any actions match the action supplied
+            if (context.getActionPrivileges() instanceof RoleBasedActionPrivileges roleBasedActionPrivileges) {
+                Set<String> actions = roleBasedActionPrivileges.flattenedActionGroups().resolve(accessLevels);
+                // a matcher to test against all patterns in `actions`
+                WildcardMatcher matcher = WildcardMatcher.from(actions, true);
+                if (matcher.test(action)) {
+                    presponse.allowed = true;
+                    log.debug("Resource {} is shared with user {}, granting access.", req.id(), user.getName());
+                } else {
+                    // TODO check why following addition doesn't reflect in the final response message and find an alternative
+                    presponse.getMissingPrivileges().add(action);
```

Comment:
I'd believe that's because the collections managed by PrivilegesEvaluatorResponse are immutable. It's a bit surprising that no exception is thrown.

Generally, I would recommend to avoid the pattern of modifying `PrivilegesEvaluatorResponse` inline. An alternative would be:

```
return PrivilegesEvaluatorResponse.insufficient(action)
```

Reply:
With move to async evaluator, I was able to call .insufficient() method.